src/WasmExecutor.cpp

```diff
@@ -226,75 +247,131 @@ class BDMalloc {
+    {
+        std::unique_ptr<CodeGen_WebAssembly> cg(new CodeGen_WebAssembly(module.target()));
+        cg->set_context(context);
+        fn_module = cg->compile(module);
+        stack_size += cg->get_requested_alloca_total();
+    }
 
-static_assert(V8_API_VERSION >= 75,
-              "Halide requires V8 v7.5 or later when compiling WITH_V8.");
+    stack_size = align_up(stack_size);
+    wdebug(1) << "Requesting stack size of " << stack_size << "\n";
 
-namespace Halide {
-namespace Internal {
-namespace {
+    std::unique_ptr<llvm::Module> llvm_module =
+        link_with_wasm_jit_runtime(&context, module.target(), std::move(fn_module));
 
-v8::Local<v8::String> NewLocalString(v8::Isolate *isolate, const char *s) {
-#if V8_API_VERSION >= 76
-    return v8::String::NewFromUtf8(isolate, s).ToLocalChecked();
-#else
-    return v8::String::NewFromUtf8(isolate, s);
+    llvm::SmallVector<char, 4096> object;
+    llvm::raw_svector_ostream object_stream(object);
+    compile_llvm_module_to_object(*llvm_module, object_stream);
+
+    // TODO: surely there's a better way that doesn't require spooling things
+    // out to temp files
+    TemporaryFile obj_file("", ".o");
+    write_entire_file(obj_file.pathname(), object.data(), object.size());
+#if WASM_DEBUG_LEVEL
+    obj_file.detach();
+    wdebug(1) << "Dumping obj_file to " << obj_file.pathname() << "\n";
 #endif
-}
 
-using namespace v8;
+    TemporaryFile wasm_output("", ".wasm");
 
-using wasm32_ptr_t = int32_t;
+    std::string lld_arg_strs[] = {
+        "HalideJITLinker",
+        // For debugging purposes:
+        // "--verbose",
+        // "-error-limit=0",
+        // "--print-gc-sections",
+        "--export=__heap_base",
+        "--allow-undefined",
+        "-zstack-size=" + std::to_string(stack_size),
+        obj_file.pathname(),
+        "--entry=" + fn_name,
+        "-o",
+        wasm_output.pathname()};
 
-const wasm32_ptr_t kMagicJitUserContextValue = -1;
+    constexpr int c = sizeof(lld_arg_strs) / sizeof(lld_arg_strs[0]);
+    const char *lld_args[c];
+    for (int i = 0; i < c; ++i) {
+        lld_args[i] = lld_arg_strs[i].c_str();
+    }
 
-#if WASM_DEBUG_LEVEL
-void print_object_properties(Isolate *isolate, const Local<Value> &v) {
-    Local<Context> context = isolate->GetCurrentContext();
-    String::Utf8Value objascii(isolate, v);
-    wdebug(0) << *objascii << "\n";
-
-    if (v->IsObject()) {
-        Local<Object> obj = v.As<Object>();
-        Local<Array> properties = obj->GetPropertyNames(context).ToLocalChecked();
-        int len = properties->Length();
-        wdebug(0) << "Number of properties = " << len << ":\n";
-        for (int i = 0; i < len; ++i) {
-            const v8::Local<v8::Value> key = properties->Get(i);
-            String::Utf8Value str(isolate, key);
-            wdebug(0) << "\t" << i + 1 << ". " << *str << "\n";
-        }
+    // lld will temporarily hijack the signal handlers to ensure that temp files get cleaned up,
```

Comment:
This sounded painful to debug.

Reply:
it wasn't the worst thing I've ever had to do, but it wasn't much fun, either